webapp/app/js/perf.js

```diff
@@ -368,34 +388,119 @@ perf.factory('PhCompare', [ '$q', '$http', 'thServiceDomain', 'PhSeries',
 }]);
 
 
+// For all those functions - 0 indicates a missing value both in input and output.
+// This is consistent with the talos data where a measurement is always positive,
+// and a missing value can end up as 0.
 perf.factory('math', [ function() {
 
-  return {
-    /**
-     * Compute the standard deviation for an array of values.
-     *
-     * @param values
-     *        An array of numbers.
-     * @param avg
-     *        Average of the values.
-     * @return a number (the standard deviation)
-     */
-    stddev: function(values, avg) {
-      if (values.length <= 1) {
-        return 0;
+  function badValues(msg) {
+    try {
+      console.log("Warning: " + msg);
+    } catch (e) { // In case we don't have console.log
+    }
+  };
+
+  // returns true if there are at least minValues values and none of them is 0
+  function isSetValid(minValues, values) {
+    if (!values || (minValues && !values.length) || values.length < minValues) {
+      badValues("Math set invalid - empty or too small:" + values);
+      return false;
+    }
+
+    for (var i = 0; i < values.length; i++) {
+      if (!values[i]) {
+        badValues("Math set invalid - includes 0: " + values);
+        return false;
       }
+    }
 
-      return Math.sqrt(
-        values.map(function (v) { return Math.pow(v - avg, 2); })
-          .reduce(function (a, b) { return a + b; }) / (values.length - 1));
-    },
+    return true;
+  }
 
-    trimFloat: function(number) {
-      if (number === undefined)
-        return 'N/A';
-      return Math.round(number * 100) / 100;
+  function percentOf(a, b) {
+    return 100 * a / b;
+  }
+
+  function geomean(values) {
+    if (!isSetValid(1, values)) {
+      return 0;
     }
-  };
+
+    var rv = 1;
+    for (var i = 0; i < values.length; i++) {
+        rv *= values[i];
+    }
+    return Math.pow(rv, 1 / values.length);
+  }
+
+  function stddev(values, avg) {
+    if (!isSetValid(2, values)) {
+      return 0;
+    }
+
+    if (!avg)
+      avg = geomean(values);
+
+    return Math.sqrt(
+      values.map(function (v) { return Math.pow(v - avg, 2); })
+        .reduce(function (a, b) { return a + b; }) / (values.length - 1));
+  }
+
+  // If a set has only one value, assume average-ish-plus sddev, which
+  // will manifest as smaller t-value the less items there are at the group
+  // (so quite small for 1 value). This default value is a parameter.
+  // C/T mean control/test group (in our case original/new data).
+  // Assumption: all the values are positive.
+  function t_test(valuesC, valuesT, stddev_default_factor) {
+      // We must have at least one value at each set
```

Comment:
nit: this function is indented 4 spaces whereas the rest of the code is 2 space; lets stick to 2 space.